src/relay/backend/vm/compiler.cc

```diff
@@ -419,164 +439,183 @@ struct VMCompiler : ExprFunctor<void(const Expr& expr)> {
     }
   }
 
-  void CompileClosure(const Function& func) {
-    // We first layout the function arguments.
-    auto inner_func = Downcast<Function>(func->body);
-
-    size_t i = 0;
-    for (auto param : inner_func->params) {
-      auto arg_register = NewRegister();
-      CHECK_EQ(i, arg_register);
-      var_register_map.insert({param, arg_register});
-      i++;
-    }
-
-    // We then assign register num to the free variables
-    for (auto param : func->params) {
-      auto arg_register = NewRegister();
-      CHECK_EQ(i, arg_register);
-      var_register_map.insert({param, arg_register});
-      i++;
-    }
-
-    // We will now process the body like normal.
-    this->VisitExpr(inner_func->body);
-  }
-
-  void Compile(const Function& func) {
-    // We need to generate code specially for lifted closures.
-    if (IsClosure(func)) {
-      CompileClosure(func);
-      return;
-    }
-
-    for (size_t i = 0; i < func->params.size(); ++i) {
-      auto arg_register = NewRegister();
-      CHECK_EQ(arg_register, i);
-      var_register_map.insert({func->params[i], arg_register});
-    }
-
-    this->VisitExpr(func->body);
-  }
+ protected:
+  /*! \brief Store the expression a variable points to. */
+  std::unordered_map<Var, Expr, NodeHash, NodeEqual> expr_map_;
+  /*! \brief Instructions in the VMFunction. */
+  std::vector<Instruction> instructions_;
+  /*! \brief Parameter names of the function. */
+  std::vector<std::string> params_;
+  /*! \brief Map from var to register number. */
+  std::unordered_map<Var, RegName, NodeHash, NodeEqual> var_register_map_;
+  /*! \brief Last used register number. */
+  size_t last_register_;
+  /*! \brief Total number of virtual registers allocated. */
+  size_t registers_num_;
+  /*! \brief Compiler engine to lower primitive functions. */
+  CompileEngine engine_;
+  /*! \brief Global shared meta data */
+  VMCompilerContext* context_;
+  /*! \brief Target devices. */
+  TargetsMap targets_;
 };
 
-void PopulatePackedFuncMap(const std::vector<LoweredFunc>& lowered_funcs,
-                           std::vector<PackedFunc>* packed_funcs) {
-  runtime::Module mod;
-  if (lowered_funcs.size() > 0) {
-    // TODO(@jroesch): we need to read target from build config
-    Target target = Target::Create("llvm");
-    if (const auto* f = runtime::Registry::Get("relay.backend.build")) {
-      mod = (*f)(tvm::Array<LoweredFunc>(lowered_funcs.begin(), lowered_funcs.end()), target);
+class VMBuildModule : public runtime::ModuleNode {
```

Comment:
Can we give this a different name? `build_module` is a general pet-peeve of mine for a compiler it has no previous meaning and doesn't clearly illustrate what this does. Module is also an incredibly overloaded word in TVM and is pretty ambiguous at this point.

Reply:
For example VMCompiler is a much better name imo.